book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+
+Using Python, JavaScript, C++ or Rust, one could define many different a data
+structure that would correspond to our idea of "relation". Most of them would be
+an object/struct that has column names and types and then a generic array of
+arrays for rows.
+
+PRQL's type system should also be able to express relations as composed from
+primitive types, but have only one idiomatic way of doing so.
+
+In practice this means that builtin types include only primitives (int, text,
+bool, float), struct (for product), enum (for sum) and list (for repeating).
+
+An SQL row would translate to struct, and a relation would translate to a list
+of structs.
+
+I would also strive for the type system to be minimal - don't differentiate
+between tuples, objects and structs. Choose one and stick to it.
+
+**Type constraints** - constrain a type with a predicate. For example, have a
+type of int64 that are equal or greater than 10. Postgres
+[does support this](https://news.ycombinator.com/item?id=34835063). The primary
+value of using constrained types would not be validation (as it is used in
+linked article), but when matching the type.
+
+Say, for example, that we have a pipeline like this:
+
+```
+derive color = switch [x => 'red', true => 'green']
+derive is_red = switch [color == 'red' => true, color == 'green' => false]
+```
+
+It should be possible to infer that `color` is of type `text`, but only when
+equal to `'red'` or `'green'`. This means that the second switch covers all
+possible cases and `is_red` cannot be `null`.
+
+## Physical layout
+
+_Logical type_ is user-facing the notion of a type that is the building block of
+the type system.
+
+_Physical layout_ is the underlying memory layout of the data represented by a
+variable.
+
+In many programming languages, physical layout of a logical type is dependent on
+the target platform. Similarly, physical layout of a PRQL logical type is
+dependent on representation of that type in target STS.
+
+```
+PTS logical type  --->  STS logical type  ---> STS physical layout
+```
+
+Note that STS types do not have a single physical layout. Postgres has a logical
+type `anyelement`, which is a super type of any data type. It can be used as a
+function parameter type, but does not have a single physical layout so it cannot
+be used in a column declaration.
+
+For now, PRQL does not define a physical layout of any type. It is not needed
+since PRQL is not used for DDL (see section "Built-in primitives") or does not
+support raw access to underlying memory.
+
+As a consequence, results of a PRQL query cannot be robustly compared across
+DBMSs, since the physical layout of the result will vary.
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
+type float
+type bool
+type text
+type char
+type null
+
+# users-defined types
+type my_int = int
```

Comment:
> @snth 
> We currently use angle brackets <> for type annotations. I have mostly seen this used in discussions in Github issues and in the [PRQL std lib](https://prql-lang.org/book/language-features/standard-library.html). To me it seems like <> acts as a "type constructor" operator (I'm not sure if that's completely correct in the most technical sense of the word) in the sense that it takes an expression and turns it into a type. I think we should try to keep that and ensure that whatever syntax we decide on stays consistent with that. What I mean by that is the following:
> 
>  -  While for function defintions we write
>     ```
>     func interp lower:0 higher x -> (x - lower) / (higher - lower)
>     ```
>     this is equivalent to
>     ```
>     let interp = lower:0 higher x -> (x - lower) / (higher - lower)
>     ```
>     but the first form is preferred for readability.
>  -  Similarly for types we could(/should) have that
>     ```
>     type number = int | float
>     ```
>     is equivalent to
>     ```
>     let number = <int | float>
>     ```
>     where again the first is preferred for readability but they are equivalent.

This is a great point: I haven't really defined how types interact with values and what `< >` means.

I think we should define a type of a variable to be "a set of all possible values of that variable". And if types are just sets, any type expression are just set expressions (e.g. `int || null` is a union of two sets).

`param <X>` would then mean "param can accept any value from set X".

I'd say that the syntax for your last example should then be:

```
type number = int || float
let number = int || float
```